### PR TITLE
core: Add OIDC options and tests

### PR DESCRIPTION
**Pull request by jtszalay**
Resolves https://github.com/mohammed90/caddy-ngrok-listener/issues/15

should be merged after #18

### CODE REVIEW

**Comment by jtszalay** [comment]
After this PR is reviewed and if all looks good I should be able to knock out OAuth and Webhook Verification immediately after as they're the same pattern! 🎉

**Comment by jtszalay** [comment]
This one is ready for review!

File: oidc.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	if o.AllowDomains != nil {
+		o.opts = append(o.opts, config.WithAllowOIDCDomain(o.AllowDomains...))
+	}
+
+	if o.Scopes != nil {
+		o.opts = append(o.opts, config.WithOIDCScope(o.Scopes...))
+	}
+
```
Comment (mohammed90) [inline oidc.go:38]: Change the check to `len(...) > 0`. See https://github.com/mohammed90/caddy-ngrok-listener/commit/8d3ac56e60cb8b5cd36ee10abb683565ee140adb for earlier unintended bug.

File: oidc.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+	if o.ClientID == "" {
+		return errors.New("oidc `client_id` cannot be empty string")
+	}
+
+	if o.ClientSecret == "" {
+		return errors.New("oidc `client_secret` cannot be empty string")
+	}
+
```
Comment (mohammed90) [inline oidc.go:50]: To be safe, should it have `strings.TrimSpace`?

**Review by mohammed90 — CHANGES_REQUESTED** [review]
Alright, let's try to pick up from where we left 😄

File: oidc.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	IssuerURL    string   `json:"issuer_url,omitempty"`
+	ClientID     string   `json:"client_id,omitempty"`
+	ClientSecret string   `json:"client_secret,omitempty"`
+	AllowEmails  []string `json:"allow_emails,omitempty"`
+	AllowDomains []string `json:"allow_domains,omitempty"`
+	Scopes       []string `json:"scopes,omitempty"`
+}
+
+func (o *oidc) Provision(caddy.Context) error {
```
[the remark this answers is not in the document]
Comment (mohammed90) [reply oidc.go:23]: Ah, I see now unexporting the method breaks the testing facade. That's alright.

**Review by mohammed90 — APPROVED** [review]
Thank you for all your work on this! I'll try to pick up 🙂 If you're still interested, I'd love to have your view on how this module progresses.

**Pull request merged**